### DISCUSSION TITLE
IRQ between StateMachine blocks

### DISCUSSION THREAD

**Discussion by andrekuehne** [micropython/micropython#11114] [RP2040 / Pico] [answered]
Is it possible to synchronize multiple PIO state machines that live on separate blocks using IRQs? The following code blinks an LED on GPIO0 and works when both state machines share the same block (e.g. _SM_BLINK_ID=0, _SM_IRQ_ID = 3 or _SM_BLINK_ID=4, _SM_IRQ_ID = 7 ) but fails once they live on different blocks  (e.g. _SM_BLINK_ID=0, _SM_IRQ_ID = 4). Is there a potential alternative to achieve the desired behavior? Thanks!

```python
from rp2 import PIO, asm_pio, StateMachine
from machine import Pin

_IRQ_NUM = const(1)
_SM_BLINK_ID = 0
_SM_IRQ_ID = 1

@asm_pio(set_init=PIO.OUT_LOW)
def blink_on_irq():
    irq(block,_IRQ_NUM)
    set(pins,0b1)
    irq(block,_IRQ_NUM)
    set(pins,0b0)
    
@asm_pio()
def timed_irq():
    pull(block) # get delay
    mov(x,osr)  # store in x
    mov(y,x)    # store in y
    wrap_target()
    irq(clear,_IRQ_NUM)
    label("delay")
    jmp(x_dec,"delay")
    mov(x,y)

sm_blink = StateMachine(_SM_BLINK_ID,blink_on_irq,set_base=Pin(0),freq=2000)
sm_irq = StateMachine(_SM_IRQ_ID,timed_irq,freq=2000)
sm_blink.active(1)
sm_irq.active(1)
sm_irq.put(500)

**Comment by rkompass** [comment, marked as answer]
Simple answer is: No.  The IRQ flags of the two PIO blocks reside in different memory areas.

**Reply by andrekuehne** [reply]
Thanks! So the only way to communicate between the SM blocks would be involving GPIO pins I assume? At least that's my current workaround - setting a pin from one SM and waiting on it on the other one.

**Reply by rkompass** [reply]
Apparently. GPIO came to my mind too. Do you have such complicated stuff running on the PIOs that one SM block does not suffice?

**Reply by andrekuehne** [reply]
I need to create four pulses of different polarity and configurable timing for an instrument control. Since the pulses can be more or less independent and potentially overlapping, the easiest solution was to build a state machine for pulses of varying pre-delay, pulse length and post delay that I instantiate four times with different parameters. The machines then wait for the same external trigger event or IRQ. The fifth machine comes into play once I want to run configurable pulse train patterns of those individual pulses using a state machine that defines the general pattern. I wanted to trigger the individual pulse playback using an IRQ, but the GPIO solution works nicely. Fortunately I still had a GPIO pin left over. 😅
